src/tools/periph/mkfatimage16.c

```diff
@@ -467,24 +467,31 @@ int main(int argc, char *argv[])
   bpb->num_heads = heads;
 
   if (bootsect_file) {
-    switch (bpb->v340.signature) {
-      case 0x29: // v400
+    if(bpb->v340.signature == 0x28 || bpb->v340.signature == 0x29) {
+      if(bpb->v340.signature == 0x29) { // v400
         memset(bpb->v400.vol_label, ' ', 11);
         memcpy(bpb->v400.vol_label, volume_label, strlen(volume_label));
         memcpy(bpb->v400.fat_type,
                p_type == P_TYPE_12BIT ? "FAT12   " : "FAT16   ", 8);
-        /* fall through */
-      case 0x28: // v340
-        bpb->v340.drive_number = 0x80;
-        bpb->v340.flags = 0;
-        bpb->v340.serial_number = 0x12345678;
-        /* fall through */
-      default: //  v331 compatible
-        if (p_sectors >= 65536) { // Only rewrite if we are large
-          bpb->v331.hidden_sectors = HIDDEN_SECTORS;
-          bpb->v331.num_sectors_large = p_sectors;
-        }
-        break;
+      }
+      // v400 && v340
+      bpb->v340.drive_number = 0x80;
+      bpb->v340.flags = 0;
+      bpb->v340.serial_number = 0x12345678;
+
+      bpb->v340.hidden_sectors = HIDDEN_SECTORS;
+      bpb->v340.num_sectors_large = (p_sectors < 65536L) ? 0 : p_sectors;
+    } else {
+      if (buffer[0x1fd] >= 0x80 && buffer[0x1fd] <= 0x83) {
+        buffer[0x1fd] = 0x80;    // drive number
+      }
+
+      if(p_sectors >= 65536) { // must be v331+ if we have FAT16B
+        bpb->v331.hidden_sectors = HIDDEN_SECTORS;
+        bpb->v331.num_sectors_large = p_sectors;
+      } else {                 // try to be compatible with v3.00
+        bpb->v300.hidden_sectors = HIDDEN_SECTORS;
+      }
     }
   } else { // Assume we are writing for Dosemu / FreeDOS so v4 BPB
     bpb->v400.hidden_sectors = HIDDEN_SECTORS;
```

Comment:
We have no struct disk in mkfatimage16, only calculated values. Actually I was toying with the idea of creating a dummy struct disk and using the geometry fixup routine from disks.c to cull some duplicate code.